core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
 import { ExpressionParserService } from
   'expressions/expression-parser.service.ts';
 
-// Exceptions that can be thrown from the evaluation of expressions.
-const ExpressionError = function(): void {};
-ExpressionError.prototype = new Error();
-ExpressionError.prototype.constructor = ExpressionError;
-
-const ExprUndefinedVarError = function(varname: string,
-    envs: Array<object>): void {
-  this.varname = varname;
-  this.envs = envs;
-};
-ExprUndefinedVarError.prototype = new ExpressionError();
-ExprUndefinedVarError.prototype.constructor = ExprUndefinedVarError;
-ExprUndefinedVarError.prototype.name = 'ExprUndefinedVarError';
-ExprUndefinedVarError.prototype.toString = function(): string {
-  return this.name + ': ' + this.varname + ' not found in ' + this.envs;
-};
-
-const ExprWrongNumArgsError = function(args: Array<number|string>,
-    expectedMin: number, expectedMax: number): void {
-  this.args = args;
-  this.expectedMin = expectedMin;
-  this.expectedMax = expectedMax;
-};
-ExprWrongNumArgsError.prototype = new ExpressionError();
-ExprWrongNumArgsError.prototype.constructor = ExprWrongNumArgsError;
-ExprWrongNumArgsError.prototype.name = 'ExprWrongNumArgsError';
-ExprWrongNumArgsError.prototype.toString = function(): string {
-  return this.name + ': {' + this.args + '} not in range [' +
+class ExpressionError extends Error {
+  constructor() {
+    super();
+    // NOTE TO DEVELOPERS: As per the recommendation of TypeScript
+    //  we need to adjust the prototype manually right after the super() call
+    //  For more information refer to this link below:
+    //  https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, ExpressionError.prototype);
```

Comment:
Can we just remove the call to super entirely? All of these constructors should have an empty body.

Reply:
@brianrodri Sorry if I'm bothering you again but if I would remove this (leaving the constructor body blank except `super()` then tests will fail again. This is why I added this line of code and the error looks like this which I can't follow after investigating it:
```
Expression type parser service should determine the correct types for the expressions FAILED
	Error: Expected true to be false.
```
```
ExprUndefinedVarError thrown
	error properties: Object({ varname: 'a', envs: [ Object({  }), Object({ +: Object({ eval: Function, getType: Function }), -: Object({ eval: Function, getType: Function }), *: Object({ eval: Function, getType: Function }), /: Object({ eval: Function, getType: Function }), %: Object({ eval: Function, getType: Function }), <=: Object({ eval: Function, getType: Function }), >=: Object({ eval: Function, getType: Function }), <: Object({ eval: Function, getType: Function }), >: Object({ eval: Function, getType: Function }), !: Object({ eval: Function, getType: Function }), ==: Object({ eval: Function, getType: Function }), !=: Object({ eval: Function, getType: Function }), &&: Object({ eval: Function, getType: Function }), ||: Object({ eval: Function, getType: Function }), if: Object({ eval: Function, getType: Function }), floor: Object({ eval: Function, getType: Function }), pow: Object({ eval: Function, getType: Function }), log: Object({ eval: Function, getType: Function }), abs: Object({ eval: Function,  ...
	    at <Jasmine>
	    at ExprUndefinedVarError.ExpressionError [as constructor] (core/templates/combined-tests.spec.js:22806:72353)
```

Reply:
@brianrodri Sorry if I'm bothering you again but if I would remove this (leaving the constructor body blank except `super()` then tests will fail again. This is why I added this line of code and the error looks like this which I can't follow after investigating it:
```
Expression type parser service should determine the correct types for the expressions FAILED
	Error: Expected true to be false.
```
```
ExprUndefinedVarError thrown
	error properties: Object({ varname: 'a', envs: [ Object({  }), Object({ +: Object({ eval: Function, getType: Function }), -: Object({ eval: Function, getType: Function }), *: Object({ eval: Function, getType: Function }), /: Object({ eval: Function, getType: Function }), %: Object({ eval: Function, getType: Function }), <=: Object({ eval: Function, getType: Function }), >=: Object({ eval: Function, getType: Function }), <: Object({ eval: Function, getType: Function }), >: Object({ eval: Function, getType: Function }), !: Object({ eval: Function, getType: Function }), ==: Object({ eval: Function, getType: Function }), !=: Object({ eval: Function, getType: Function }), &&: Object({ eval: Function, getType: Function }), ||: Object({ eval: Function, getType: Function }), if: Object({ eval: Function, getType: Function }), floor: Object({ eval: Function, getType: Function }), pow: Object({ eval: Function, getType: Function }), log: Object({ eval: Function, getType: Function }), abs: Object({ eval: Function,  ...
	    at <Jasmine>
	    at ExprUndefinedVarError.ExpressionError [as constructor] (core/templates/combined-tests.spec.js:22806:72353)
```

Reply:
Remove super as well, it isn't necessary (not sure why you added it in the first place?)

Reply:
@brianrodri If I would leave the body empty the linter will yell again:
`Constructor for derived classes must contain a 'super' call`

Reply:
Ah I didn't notice you were subclassing from `Error` in my first pass. OK, let's try to use `new.target.prototype` instead of hard-coding values: https://stackoverflow.com/a/58417721/4859885

Reply:
@brianrodri Great! It's exactly what I was looking for thanks 👍 
Should I add a comment to explain with the link above just like the previous?

Reply:
Yes please, that would be super helpful (no pun intended)